components/common/src/lib.rs

```diff
@@ -146,7 +146,171 @@ impl FeatureFlag {
+        {
+            error!("No heartbeat from {} in {} seconds; deadlock likely",
+                   name.unwrap_or_else(|| { "unnamed thread".to_string() }),
+                   last_heartbeat.elapsed().as_secs());
+        }
+    }
+
+    fn threads_missing_heartbeat(heartbeats: &HeartbeatMap,
+                                 threshold: Duration)
+                                 -> Vec<NameAndLastHeartbeat> {
+        heartbeats.iter()
+                  .filter_map(|(thread_id, (thread_name, last_heartbeat))| {
+                      let time_since_last_heartbeat = last_heartbeat.elapsed();
+                      trace!("{:?} {:?} last heartbeat: {:?} ago",
+                             thread_id,
+                             thread_name,
+                             time_since_last_heartbeat);
+                      if time_since_last_heartbeat < threshold {
+                          None
+                      } else {
+                          Some((thread_name.clone(), last_heartbeat.clone()))
+                      }
+                  })
+                  .collect::<Vec<_>>()
+    }
+
+    #[cfg(test)]
+    mod test {
+        use super::*;
+
+        const TEST_THRESHOLD: Duration = Duration::from_millis(10);
+
+        #[test]
+        fn no_tracking_without_mark_thread_alive() {
+            let heartbeats = HashMap::new();
+            thread::spawn(|| {}).join().unwrap();
+            thread::sleep(TEST_THRESHOLD * 2);
+            assert!(threads_missing_heartbeat(&heartbeats, TEST_THRESHOLD).is_empty());
+        }
+
+        #[test]
+        fn one_dead_thread() {
+            lazy_static! {
+                static ref HEARTBEATS: Mutex<HeartbeatMap> = Default::default();
+            }
+            thread::spawn(move || mark_thread_alive_impl(&mut HEARTBEATS.lock().unwrap())).join()
+                                                                                          .unwrap();
+            thread::sleep(TEST_THRESHOLD * 2);
+            assert_eq!(threads_missing_heartbeat(&HEARTBEATS.lock().unwrap(), TEST_THRESHOLD).len(),
+                       1);
+        }
+
+        #[test]
+        fn one_dead_one_alive() {
+            lazy_static! {
+                static ref HEARTBEATS: Mutex<HeartbeatMap> = Default::default();
+            }
+
+            let dead_thread_name = "expected-dead".to_string();
+
+            thread::Builder::new().name(dead_thread_name.clone())
+                                  .spawn(move || {
+                                      mark_thread_alive_impl(&mut HEARTBEATS.lock().unwrap())
+                                  })
+                                  .unwrap()
+                                  .join()
+                                  .unwrap();
+            thread::spawn(move || {
+                loop {
+                    mark_thread_alive_impl(&mut HEARTBEATS.lock().unwrap());
+                    thread::sleep(TEST_THRESHOLD / 2)
+                }
+            });
```

Comment:
Do you need to clean this thread up? Or does it end up killing itself at the end of the function when `HEARTBEATS` gets dropped?

Also, does the `HEARTBEATS` mutex in these tests need to be in a `lazy_static!` block in the first place?

Reply:
This thread should get cleaned up when the test runner exits. I suppose it would be better to have it shut down right at the end of this particular test. I'll see how much messier that makes things.

Reply:
See https://github.com/habitat-sh/habitat/pull/6565/commits/419d60dbe29be4c931108a60fd36f1f51574fd6a#diff-ce8ce0bdcb96f29e33454e96fcc78dafR275